src/rand.c

```diff
@@ -0,0 +1,231 @@
+#endif
+
+static uint64_t state[4];
+static git_mutex state_lock;
+
+/*
+ * On 32-bit systems, we want to put inputs to the seed in the high
+ * and low parts of the 64-bit seed, so we shift some into place.
+ */
+#ifdef GIT_ARCH_32
+# define SEED_SHIFT 32
+#else
+# define SEED_SHIFT 0
+#endif
+
+#if defined(GIT_WIN32)
+GIT_INLINE(int) getseed(uint64_t *seed)
+{
+	HCRYPTPROV provider;
+	SYSTEMTIME systemtime;
+	FILETIME filetime, idletime, kerneltime, usertime;
+
+	if (CryptAcquireContext(&provider, 0, 0, PROV_RSA_FULL,
+	                        CRYPT_VERIFYCONTEXT|CRYPT_SILENT)) {
+		BOOL success = CryptGenRandom(provider, sizeof(uint64_t), (void *)seed);
+		CryptReleaseContext(provider, 0);
+
+		if (success)
+			return 0;
+	}
+
+	GetSystemTime(&systemtime);
+	if (!SystemTimeToFileTime(&systemtime, &filetime)) {
+		git_error_set(GIT_ERROR_OS, "could not get time for random seed");
+		return -1;
+	}
+
+	/* Fall-through: generate a seed from the time and system state */
+	*seed = 0;
+	*seed |= ((uint64_t)filetime.dwLowDateTime << 32);
+	*seed |= ((uint64_t)filetime.dwHighDateTime);
+
+	*seed ^= ((uint64_t)GetCurrentProcessId() << 32);
+	*seed ^= ((uint64_t)GetCurrentThreadId());
+
+	GetSystemTimes(&idletime, &kerneltime, &usertime);
+	*seed |= ((uint64_t)idletime.dwLowDateTime);
+	*seed |= ((uint64_t)idletime.dwHighDateTime << 32);
+	*seed |= ((uint64_t)kerneltime.dwLowDateTime << 32);
+	*seed |= ((uint64_t)kerneltime.dwHighDateTime);
+	*seed |= ((uint64_t)usertime.dwLowDateTime);
+	*seed |= ((uint64_t)usertime.dwHighDateTime << 32);
+
+	/* Mix in the addresses of some functions and variables */
+	*seed ^= (((uint64_t)((uintptr_t)getseed)) << SEED_SHIFT);
+	*seed ^= (((uint64_t)((uintptr_t)seed)));
+	*seed ^= (((uint64_t)((uintptr_t)printf)) << SEED_SHIFT);
+	*seed ^= (((uint64_t)((uintptr_t)&errno)));
+
+	*seed ^= ((uint64_t)git__timer());
+
+	return 0;
+}
+
+#else
+
+GIT_INLINE(int) getseed(uint64_t *seed)
+{
+	struct timeval tv;
+	double loadavg[3];
+	int fd;
+
+# if defined(GIT_RAND_GETENTROPY)
+	GIT_UNUSED((fd = 0));
+
+	if (getentropy(seed, sizeof(uint64_t)) == 0)
+		return 0;
+# else
+	/*
+	 * Try to read from /dev/urandom; most modern systems will have
+	 * this, but we may be chrooted, etc, so it's not a fatal error
+	 */
+	if ((fd = open("/dev/urandom", O_RDONLY)) >= 0) {
+		ssize_t ret = read(fd, seed, sizeof(uint64_t));
+		close(fd);
+
+		if (ret == (ssize_t)sizeof(uint64_t))
+			return 0;
+	}
+# endif
+
+	/* Fall-through: generate a seed from the time and system state */
+	*seed = 0;
+
+	if (gettimeofday(&tv, NULL) < 0) {
+		git_error_set(GIT_ERROR_OS, "could get time for random seed");
+		return -1;
+	}
+
+	*seed = 0;
+	*seed |= ((uint64_t)tv.tv_usec << 40);
+	*seed |= ((uint64_t)tv.tv_sec);
+
+	*seed ^= ((uint64_t)getpid() << 32);
+	*seed ^= ((uint64_t)getpgid(0));
+	*seed ^= ((uint64_t)getppid() << 32);
+	*seed ^= ((uint64_t)getsid(0));
+	*seed ^= ((uint64_t)getuid() << 32);
+	*seed ^= ((uint64_t)getgid());
+
```

Comment:
I think there is a missing call to getloadavg(loadavg,3) here, unfortunately it isn't a POSIX function so we might need to check if it exists.